addon-chrome/ChromeActivityStreams.js

```diff
@@ -0,0 +1,243 @@
+          break;
+        case "NOTIFY_BLOCK_URL":
+          this._blockUrl(action);
+          break;
+        case "NOTIFY_UNBLOCK_ALL":
+          this._unblockAll(action);
+          break;
+        case "NOTIFY_OPEN_WINDOW":
+          this._openNewWindow(action);
+          break;
+        case "SEARCH_SUGGESTIONS_REQUEST":
+          this._searchSuggestions(action);
+          break;
+        case "NOTIFY_PERFORM_SEARCH":
+          this._performSearch(action);
+          break;
+        case "SEARCH_STATE_REQUEST":
+          this._searchState();
+          break;
+        case "SEARCH_UISTRINGS_REQUEST":
+          this._searchUIStrings();
+          break;
+      }
+    }, false);
+  }
+
+  _setupChromeListeners() {
+    chrome.history.onVisited.addListener((result) => {
+      const row = ChromePlacesProvider.transformHistory(result);
+      ChromePlacesProvider.addHistory(row).then((result) => {
+        ChromePlacesProvider.getHistory().then((histories) => {
+          dispatch({
+            type: "RECENT_LINKS_RESPONSE",
+            data: histories
+          });
+        });
+      });
+    });
+
+    chrome.history.onVisitRemoved.addListener((result) => {
+      result.urls.forEach((url) => {
+        ChromePlacesProvider.removeHistory(url);
+        dispatch({
+          type: "NOTIFY_HISTORY_DELETE",
+          data: url
+        });
+      });
+    });
+
+    chrome.bookmarks.onCreated.addListener((id, result) => {
+      const isFolder = !result.url;
+      if (isFolder) return;
+      const row = ChromePlacesProvider.transformBookmark(result);
+      ChromePlacesProvider.addBookmark(row);
+      dispatch({
+        type: "RECENT_BOOKMARKS_RESPONSE",
+        data: [row],
+        meta: {prepend: true}
+      });
+    });
+
+    chrome.bookmarks.onRemoved.addListener((result) => {
+      ChromePlacesProvider.removeBookmark(result);
+      dispatch({
+        type: "NOTIFY_BOOKMARK_DELETE",
+        data: result
+      });
+    });
+  }
+
+  _topFrecentSites(action) {
+    ChromePlacesProvider.getHistory().then((histories) => {
+      // https://dxr.mozilla.org/mozilla-central/source/mobile/android/base/java/org/mozilla/gecko/db/BrowserContract.java#124
+      // numVisits * max(1, 100 * 225 / (age*age + 225))
+      const rows = histories
+        .filter((hist) => !/google/.test(hist.url))
+        .map((hist) => {
+          const microsecondsPerDay = 86400000000;
+          const age = (new Date().getTime() - hist.lastVisitDate) / microsecondsPerDay;
+          return Object.assign(hist, {frencency: hist.visitCount * Math.max(1, 100 * 225 / (age * age + 225))});
+        })
+        .sort((a, b) => {
+          if (a.frencency > b.frencency) {
+            return -1;
+          }
+          if (a.frencency < b.frencency) {
+            return 1;
+          }
+          return 0;
+        });
+
+      dispatch({type: "TOP_FRECENT_SITES_RESPONSE", data:rows});
+    });
+  }
+
+  _recentBookmarks(action) {
```

Comment:
The "document your parameters" is the minimum. Also consider asserting that you've received what you expected to receive, and enforcing an interface. This will help a lot of users of these functions (activity stream app) change how they're calling these functions. You don't want this code to fail in subtle ways, this will result in hard-to-trace bugs and generally brittle behaviour.